memiavl/tree.go

```diff
@@ -52,16 +52,29 @@ func NewFromSnapshot(snapshot *Snapshot) *Tree {
 	}
 }
 
-func (t *Tree) Set(key, value []byte) {
+// ApplyChangeSet apply the change set of a whole version, and update hashes.
+func (t *Tree) ApplyChangeSet(changeSet *ChangeSet, updateHash bool) ([]byte, int64, error) {
+	for _, pair := range changeSet.Pairs {
+		if pair.Delete {
+			t.remove(pair.Key)
+		} else {
+			t.set(pair.Key, pair.Value)
+		}
+	}
+
+	return t.saveVersion(updateHash)
+}
+
+func (t *Tree) set(key, value []byte) {
 	t.root, _ = setRecursive(t.root, key, value, t.version+1)
 }
 
-func (t *Tree) Remove(key []byte) {
+func (t *Tree) remove(key []byte) {
```

Comment:
shouldn't `remove` and `set` be exported for users to have access to?

Reply:
I'll count on the `KVStore` wrapper to provide those APIs, in sdk's usage, the whole block's changes are written at once anyway.

Reply:
we have another invariant to maintain for the `set/remove` calls, they must happen in the same order as the keys.